light-rpc/src/main.rs

```diff
@@ -1,4 +1,503 @@
+            let finalized_slot = finalized_slot.clone();
+            let join_handle = Builder::new()
+                .name("thread working on confirmation block".to_string())
+                .spawn(move || {
+                    Self::process_block(
+                        receiver_confirmed,
+                        signature_status,
+                        CommitmentLevel::Confirmed,
+                        confirmation_slot,
+                    );
+                })
+                .unwrap();
+            joinables.push(join_handle);
+        }
+        // start thread which will listen to finalized blocks
+        {
+            let signature_status = signature_status.clone();
+            let confirmation_slot = confirmation_slot.clone();
+            let finalized_slot = finalized_slot.clone();
+            let join_handle = Builder::new()
+                .name("thread working on finalized block".to_string())
+                .spawn(move || {
+                    Self::process_block(
+                        receiver_finalized,
+                        signature_status,
+                        CommitmentLevel::Finalized,
+                        finalized_slot,
+                    );
+                })
+                .unwrap();
+            joinables.push(join_handle);
+        }
+
+        LightRpcRequestProcessor {
+            rpc_client,
+            tpu_client,
+            last_valid_block_height: 0,
+            ws_url: websocket_url.clone(),
+            current_slot,
+            signature_status,
+            connection_cache: connection_cache,
+            subscribed_clients: Arc::new(vec![client_confirmed, client_finalized]),
+            joinables: Arc::new(joinables),
+            confirmation_slot: confirmation_slot,
+            finalized_slot: finalized_slot,
+        }
+    }
+
+    fn process_block(
+        reciever: Receiver<RpcResponse<RpcBlockUpdate>>,
+        signature_status: Arc<RwLock<HashMap<String, Option<CommitmentLevel>>>>,
+        commitment: CommitmentLevel,
+        updated_slot: Arc<AtomicU64>,
+    ) {
+        let mut last_update_slot: u64 = 0;
+        loop {
+            let block_data = reciever.recv();
+
+            match block_data {
+                Ok(data) => {
+                    let blockUpdate = data.value;
+                    last_update_slot = blockUpdate.slot;
+                    updated_slot.swap(last_update_slot, Ordering::Relaxed);
+                    if let Some(block) = blockUpdate.block {
+                        if let Some(signatures) = block.signatures {
+                            let mut lock = signature_status.write().unwrap();
+                            for signature in signatures {
+                                if lock.contains_key(&signature) {
+                                    println!("found signature {}", signature);
+                                    lock.insert(signature, Some(commitment));
```

Comment:
`process_block` should be side-effect free if possible. so maybe the signatures could be stored in a local structure and returned altogether from `process_block`. TBH I did not think too hard about the implications!